### DISCUSSION TITLE
Security concern / bug when model properties are casted

### DISCUSSION THREAD

**Discussion by AdamEsterle** [livewire/livewire#3400] [7. Bugs (archived)]
SUMMARY:
Livewire will cast model properties BEFORE validating data and THEN save the model with RAW USER INPUT (so it never actually got validated)

REPRODUCE:
1. Cast a value on the User model to integer (or whatever cast) (`protected $casts = ['field'=>'integer'];`)
2. Make a component like this
```
class UserForm extends Component
{
    public User $user;

    protected $rules = [
        'user.field' => ['required', 'integer'],
    ];

    public function update()
    {
        $this->validate();
        $this->user->save();
    }
}
```
3. wire:model="user.field" on an input - set the value to 10.5
4. call the update method

WHAT WILL HAPPEN
1. `$this->validate()`
will FIRST cast 10.5 to an integer (10) and THEN validate
(this is because Arr:get() in the validator will convert the User model to an array which then the properties are automatically casted by Laravel)
2.`$this->user->save()`
will then get run! When it should have failed validation. FURTHERMORE, it will save the field with a value of 10.5 - the PRE-CASTED, RAW USER INPUT, VALUE

--This is the method that will get the attribute (with a value of 10.5) and return it as 10 since it is casted when the Model is converted to an array
```
Illuminate/Validation/Validator.php

 /**
     * Get the value of a given attribute.
     *
     * @param  string  $attribute
     * @return mixed
     */
    protected function getValue($attribute)
    {
        return Arr::get($this->data, $attribute);
    }
```

**Comment by terry-greenlaw** [comment]
I'm experiencing the same issue with empty dates. I tried to fix it using the updatedXXX($value) function on the LiveWire component in order to set the date to Carbon::make(null) when $value is empty. But during the save() it's still passing the raw input empty string ("") to Eloquent, which throws an error.
I've tried from the Model side, adjusting $casts, $dates, and mutators on the date column, but I haven't been able to disrupt the save from pushing the raw input empty string to Eloquent.

**Reply by rubendn** [reply]
I had a similar issue and I'm using the following on the update function.  Also tried casts and mutators but this was easier.

$this->editing->date_last_contact = empty($this->editing->date_last_contact) ? NULL : $this->editing->date_last_contact;

This is after $this->validate() and before $this->editing->save();

I'm using date|nullable in the validation rule and it picks up non-date, non-null values but without the line above, when they leave it blank (and it successfully validates), it causes an error upon save like you mentioned.

**Reply by eliasmaia** [reply]
@terry-greenlaw same here!

**Comment by AdamEsterle** [comment]
A workaround
```
public function setStartDateAttribute($value)
    {
        if (is_null($value) || $value == '') {
            $this->attributes['start_date'] = null;

            return;
        }

        try {
            $this->attributes['start_date'] = Carbon::parse($value)->format('Y-m-d');
        } catch (InvalidFormatException $e) {
            $this->attributes['start_date'] = null;
        }
    }
```
